Receive enhanced locations without navigating

kmadsen opened the pull request:
<!-- ⚠️ TEMPLATE ⚠️ -->
<!-- Template for GitHub PR descriptions. Use it as a guide on how to describe your work. Feel free to remove any section when you're opening a PR if you think it does not apply for your committed changes. -->

## Description

Breaking up this large refactor into smaller pieces for better review: https://github.com/mapbox/mapbox-navigation-android/pull/3191

All examples include an interaction with the LocationEngine, to provide current location before starting to navigate. This change moves that logic into core navigation. This will make it easier to remove LocationEngine from the examples.

This change should not change any existing behavior. One functional change in behavior to the SimpleNavigationActivity, is that it will always be map matching 📌  🗺️ 

- [ ] I have added any issue links
- [ ] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [ ] I have added the appropriate milestone and project boards

## Testing

Please describe the manual tests that you ran to verify your changes

- [ ] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR
- [ ] We might need to update / push `api/current.txt` files after running `$> make 1.0-core-update-api` (Core) / `$> make 1.0-ui-update-api` (UI) if there are changes / errors we're 🆗 with (e.g. `AddedMethod` changes are marked as errors but don't break SemVer) 🚀 If there are SemVer breaking changes add the `SEMVER` label. See [Metalava](https://github.com/mapbox/mapbox-navigation-android/blob/master/docs/metalava.md) docs
<!-- - [ ] I have added an `Activity` example in the test app showing the new feature implemented (where applicable) -->
<!-- - [ ] I have made corresponding changes to the documentation (where applicable) -->
<!-- - [ ] Any changes to strings have been published to our translation tool (where applicable) -->
<!-- - [ ] Publish `testapp` in Google Play `internal` test track (where applicable) -->

kmadsen commented:
The current path to get map-matched locations also updates the notification. And then it becomes stuck 

<img width="182" alt="Screen Shot 2020-06-25 at 6 31 17 PM" src="https://user-images.githubusercontent.com/3021882/85811135-10f4e280-b712-11ea-9ffb-b72fc1cb9f8b.png">

kmadsen commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/TripSession.kt:
```diff
@@ -21,8 +21,11 @@ internal interface TripSession {
 
-    fun start()
-    fun stop()
+    fun startTripSession()
+    fun stopTripSession()
+    fun startLocationUpdates()
+    fun stopLocationUpdates()
+    fun refresh(locationEngineRequest: LocationEngineRequest)
```
@LukasPaczos what do you think about this interface?

LukasPaczos commented:
Like I mentioned in https://github.com/mapbox/mapbox-navigation-android/pull/3191#discussion_r446116733, I'm still not a fan of this solution. If I understand correctly, the major purpose of this PR is being able to request and receive location updates without having the foreground service running, is that right? We've been discussing this topic in the past and settled on not exposing this option and complicating the APIs. Developers can always go ahead and use a standalone `LocationEngine` instance (either create one themselves or fetch from `MapboxNavigation`) to request updates outside of the Nav SDK context. Is there a stronger case now for offering both?

kmadsen commented:
> If I understand correctly, the major purpose of this PR is being able to request and receive location updates without having the foreground service running, is that right? 

The main purpose of the change is to simplify the required interaction with the LocationEngine to get working examples using MapboxNavigation. Today, clients need to build a solution for receiving the first location and they don't get to choose if it's map matched or from GPS. Clients also need to understand that internally, MapboxSession will be requesting location updates from the LocationEngine - meaning you need to stop your location engine manually. Decoupling these dependencies means you can hook up LocationEngine requests to the Android lifecycle rather than putting it in trip session state. 

> Developers can always go ahead and use a standalone LocationEngine instance (either create one themselves or fetch from MapboxNavigation) to request updates outside of the Nav SDK context. Is there a stronger case now for offering both?

Do the examples represent the simplest usage of LocationEngine? Considering that this solution is addressing all of these issues, is this not a strong case for offering both?
1. Create custom logic for [receiving the first location](https://github.com/mapbox/mapbox-navigation-android/blob/edf810e87f00667e250d3cc1ef81eb48776fbe7f/examples/src/main/java/com/mapbox/navigation/examples/core/SimpleMapboxNavigationKt.kt#L462) from the LocationEngine
1. Use a [WeakReference](https://github.com/mapbox/mapbox-navigation-android/blob/edf810e87f00667e250d3cc1ef81eb48776fbe7f/examples/src/main/java/com/mapbox/navigation/examples/core/SimpleMapboxNavigationKt.kt#L465) to prevent memory leaks
1. Start and stop location updates [depending on MapboxSession state](https://github.com/mapbox/mapbox-navigation-android/blob/edf810e87f00667e250d3cc1ef81eb48776fbe7f/examples/src/main/java/com/mapbox/navigation/examples/core/SimpleMapboxNavigationKt.kt#L341-L345) rather than Android lifecycle events. This conflicts with Android recommended practices for [requesting location updates using onResume and onPause](https://developer.android.com/training/location/request-updates)
1. Make the first location be a map matched location

> We've been discussing this topic in the past and settled on not exposing this option and complicating the APIs

May I join this discussion? So far I've only had the conversation internally with @Guardiola31337. LocationEngine management is a common complaint among clients, I've also found it difficult because of the points above.

kmadsen commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSession.kt:
```diff
@@ -149,43 +148,54 @@ class MapboxTripSession(
+                val currentDate = Date()
+                navigator.updateLocation(rawLocation, currentDate)
+                val status = navigator.getStatus(currentDate)
+                updateEnhancedLocation(status.enhancedLocation, status.keyPoints)
```
interesting behavior difference between these location updates and the trip session. these location updates do not have the status polling loop, they only get status from the raw location.

it may be required to have a polling status loop to achieve good results.

kmadsen replied:
left the status polling loop out because this has been a discussion point. we've considered removing it, having a way to receive location updates without it, would give the discussion results to look at.

Guardiola31337 commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSession.kt:
```diff
@@ -144,43 +143,57 @@ class MapboxTripSession(
+    override fun startLocationUpdates() {
+        rawLocationProvider.requestLocationUpdates { rawLocation ->
+            locationObservers.forEach { it.onRawLocationChanged(rawLocation) }
+            CoroutineScope(Dispatchers.IO).launch {
```
Curious why not using `ioJobController` 🤔?

Guardiola31337 commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSession.kt:
```diff
@@ -436,10 +434,11 @@ class MapboxTripSession(
-        locationObservers.forEach { it.onEnhancedLocationChanged(location, keyPoints) }
-    }
+    private suspend fun updateEnhancedLocation(location: Location, keyPoints: List<Location>) =
+        withContext(Dispatchers.Main) {
```
Same as above but `mainJobController`

kmadsen commented:
Concluded in an internal sync that we'll hook the start/stop trip sessions to the resume/pause activity events for the purpose of simplifying the examples

kmadsen commented:
Closing this as we're exploring different approaches for solving this. I've tried to implement the new proposed approach but ran into a larger change issue. Going to track that here: https://github.com/mapbox/mapbox-navigation-android/issues/3275
